File: cli/internal/client/processes.go
package client

import (
	"context"
	"net/url"
	"strings"
)

// ListProcesses returns currently running processes for the active profile.
func (c *Client) ListProcesses(ctx context.Context) ([]map[string]any, error) {
	var resp struct {
		Processes []map[string]any `json:"processes"`
	}
	if err := c.GetJSON(ctx, "/api/processes", &resp); err != nil {
		return nil, err
	}
	return resp.Processes, nil
}

// GetProcess returns details for a single process.
func (c *Client) GetProcess(ctx context.Context, pid string) (map[string]any, error) {
	var out map[string]any
	if err := c.GetJSON(ctx, "/api/processes/"+url.PathEscape(pid), &out); err != nil {
		return nil, err
	}
	return out, nil
}

// StopProcess terminates a process.
func (c *Client) StopProcess(ctx context.Context, pid string) error {
	return c.PostJSON(ctx, "/api/processes/"+url.PathEscape(pid)+"/stop", nil, nil)
}

// ProcessesStreamPath returns the SSE stream path for live process updates.
func (c *Client) ProcessesStreamPath() string { return "/api/processes/stream" }

// SendProcessStdin writes input to a running process. Either input_text (a
// raw string written verbatim, with the optional line_ending appended) or
// keys (a list of named keys like "Enter", "Up") may be supplied. The caller
// chooses which by populating only the relevant fields in body.
func (c *Client) SendProcessStdin(ctx context.Context, pid string, body map[string]any) (map[string]any, error) {
	var out map[string]any
	if err := c.PostJSON(ctx, "/api/processes/"+url.PathEscape(pid)+"/stdin", body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// ResizeProcessPTY sends a PTY resize for a process.
func (c *Client) ResizeProcessPTY(ctx context.Context, pid string, cols, rows int) error {
	body := map[string]int{"cols": cols, "rows": rows}
	return c.PostJSON(ctx, "/api/processes/"+url.PathEscape(pid)+"/resize", body, nil)
}

// ProcessWebSocketURL returns the ws:// (or wss://) URL for the live PTY
// stream of a process. Bearer auth is passed via Sec-WebSocket-Protocol
// (the WebSocket handshake doesn't accept Authorization headers from
// browsers, and the server matches that protocol for parity).
func (c *Client) ProcessWebSocketURL(pid string) string {
	base := c.cfg.Server
	if strings.HasPrefix(base, "https://") {
		base = "wss://" + strings.TrimPrefix(base, "https://")
	} else if strings.HasPrefix(base, "http://") {
		base = "ws://" + strings.TrimPrefix(base, "http://")
	}
	return base + "/api/processes/" + url.PathEscape(pid) + "/ws"
}

// ListAutostartProcesses returns the active profile's autostart registrations.
func (c *Client) ListAutostartProcesses(ctx context.Context) ([]map[string]any, error) {
	var resp struct {
		Autostart []map[string]any `json:"autostart"`
	}
	if err := c.GetJSON(ctx, "/api/autostart-processes", &resp); err != nil {
		return nil, err
	}
	return resp.Autostart, nil
}

// CreateAutostartFromProcess registers a live process as autostart so it
// re-launches at server boot. force=true bypasses the duplicate check.
func (c *Client) CreateAutostartFromProcess(ctx context.Context, pid string, force bool) (map[string]any, error) {
	body := map[string]any{"process_id": pid}
	if force {
		body["force"] = true
	}
	var out map[string]any
	if err := c.PostJSON(ctx, "/api/autostart-processes", body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// DeleteAutostartProcess removes an autostart registration.
func (c *Client) DeleteAutostartProcess(ctx context.Context, autostartID string) error {
	return c.Delete(ctx, "/api/autostart-processes/"+url.PathEscape(autostartID), nil)
}

// RunAutostartProcess immediately spawns the command from an autostart row.
// Returns {process_id} on success.
func (c *Client) RunAutostartProcess(ctx context.Context, autostartID string) (map[string]any, error) {
	var out map[string]any
	if err := c.PostJSON(ctx, "/api/autostart-processes/"+url.PathEscape(autostartID)+"/run", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

File: app/cli/client/processes.py
"""Process endpoints — `/api/processes*` and `/api/autostart-processes*`.

Mirrors `cli/internal/client/processes.go`. The PTY WebSocket attach itself
is handled in Phase 4 (`opa proc attach`); this module exposes the URL helper
so that command can build the ws:// URL.
"""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from app.cli.client._base import Client


async def list_processes(client: Client) -> list[dict[str, Any]]:
    resp = await client.get_json("/api/processes")
    if isinstance(resp, dict) and isinstance(resp.get("processes"), list):
        return [p for p in resp["processes"] if isinstance(p, dict)]
    return []


async def get_process(client: Client, pid: str) -> dict[str, Any]:
    resp = await client.get_json(f"/api/processes/{quote(pid, safe='')}")
    return resp if isinstance(resp, dict) else {}


async def stop_process(client: Client, pid: str) -> None:
    await client.post_json(f"/api/processes/{quote(pid, safe='')}/stop")


async def send_process_stdin(
    client: Client,
    pid: str,
    body: dict[str, Any],
) -> dict[str, Any]:
    resp = await client.post_json(
        f"/api/processes/{quote(pid, safe='')}/stdin",
        body,
    )
    return resp if isinstance(resp, dict) else {}


async def resize_process_pty(
    client: Client,
    pid: str,
    cols: int,
    rows: int,
) -> None:
    await client.post_json(
        f"/api/processes/{quote(pid, safe='')}/resize",
        {"cols": cols, "rows": rows},
    )


def processes_stream_path() -> str:
    return "/api/processes/stream"


def process_websocket_path(pid: str) -> str:
    """Path used by `connect_ws` to build the full ws:// URL for a PTY attach."""
    return f"/api/processes/{quote(pid, safe='')}/ws"


# ── autostart ─────────────────────────────────────────────────────────────

async def list_autostart_processes(client: Client) -> list[dict[str, Any]]:
    resp = await client.get_json("/api/autostart-processes")
    if isinstance(resp, dict) and isinstance(resp.get("autostart"), list):
        return [a for a in resp["autostart"] if isinstance(a, dict)]
    return []


async def create_autostart_from_process(
    client: Client,
    pid: str,
    force: bool = False,
) -> dict[str, Any]:
    body: dict[str, Any] = {"process_id": pid}
    if force:
        body["force"] = True
    resp = await client.post_json("/api/autostart-processes", body)
    return resp if isinstance(resp, dict) else {}


async def delete_autostart_process(client: Client, autostart_id: str) -> None:
    await client.delete(f"/api/autostart-processes/{quote(autostart_id, safe='')}")


async def run_autostart_process(
    client: Client,
    autostart_id: str,
) -> dict[str, Any]:
    resp = await client.post_json(
        f"/api/autostart-processes/{quote(autostart_id, safe='')}/run"
    )
    return resp if isinstance(resp, dict) else {}
